> Thx, looks much cleaner in my opinion

ok :)

I am now getting an error when compiling. Do you know how to fix this?

[960/1448] Generating GPUReconstructionOCL2Code.bc
FAILED: ....
/data/mkleiner/alice/sw/SOURCES/O2/debug_streamer/0/GPU/GPUTracking/dEdx/GPUdEdx.h:190:27: error: cannot initialize a parameter of type 'const __generic char *' with an lvalue of type 'const __constant char [11]'
    mStreamer.setStreamer("debug_dedx", "UPDATE");

/data/mkleiner/alice/sw/SOURCES/O2/debug_streamer/0/Common/Utils/include/CommonUtils/DebugStreamer.h:131:38: note: passing argument to parameter here
  GPUd() void setStreamer(const char*, const char*){};

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,6 +75,11 @@ if (DPL_ENABLE_TRACING)
   add_definitions(-DDPL_ENABLE_TRACING)
 endif()
 
+if(DEFINED ENV{ENABLE_DEBUG_STREAMER})
+  add_definitions(-DDEBUG_STREAMER)
+  MESSAGE(STATUS "Enable debug streamer")
+endif()
+
 # include macros and functions that are used in the following subdirectories'
 # CMakeLists.txt
 include(O2AddExecutable)

--- a/Common/Utils/CMakeLists.txt
+++ b/Common/Utils/CMakeLists.txt
@@ -23,6 +23,7 @@ o2_add_library(CommonUtils
                        src/BoostHistogramUtils.cxx
                        src/NameConf.cxx
                        src/IRFrameSelector.cxx
+                       src/DebugStreamer.cxx
                PUBLIC_LINK_LIBRARIES ROOT::Hist ROOT::Tree Boost::iostreams O2::CommonDataFormat O2::Headers
                                      FairLogger::FairLogger)
 
@@ -45,7 +46,8 @@ o2_target_root_dictionary(CommonUtils
                                   include/CommonUtils/VerbosityConfig.h
                                   include/CommonUtils/FileFetcher.h
                                   include/CommonUtils/NameConf.h
-                                  include/CommonUtils/IRFrameSelector.h)
+                                  include/CommonUtils/IRFrameSelector.h
+                                  include/CommonUtils/DebugStreamer.h)
 
 o2_add_test(TreeStream
             COMPONENT_NAME CommonUtils

--- a/Common/Utils/include/CommonUtils/DebugStreamer.h
+++ b/Common/Utils/include/CommonUtils/DebugStreamer.h
@@ -1,0 +1,139 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file DebugStreamer.h
+/// \brief Definition of class for writing debug informations
+/// \author Matthias Kleiner <mkleiner@ikf.uni-frankfurt.de>
+
+#ifndef ALICEO2_TPC_DEBUGSTREAMER_H_
+#define ALICEO2_TPC_DEBUGSTREAMER_H_
+
+#include "GPUCommonDef.h"
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE)
+#include "CommonUtils/ConfigurableParamHelper.h"
+#if defined(DEBUG_STREAMER)
+#include "CommonUtils/TreeStreamRedirector.h"
+#endif
+#endif
+
+namespace o2::utils
+{
+
+/// struct defining the flags which can be used to check if a certain debug streamer is used
+enum StreamFlags {
+  streamdEdx = 1 << 0, ///< stream corrections and cluster properties used for the dE/dx
+};
+
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE)
+
+inline StreamFlags operator|(StreamFlags a, StreamFlags b)
+{
+  return static_cast<StreamFlags>(static_cast<int>(a) | static_cast<int>(b));
+}
+inline StreamFlags operator&(StreamFlags a, StreamFlags b) { return static_cast<StreamFlags>(static_cast<int>(a) & static_cast<int>(b)); }
+inline StreamFlags operator~(StreamFlags a) { return static_cast<StreamFlags>(~static_cast<int>(a)); }
+
+/// struct for setting and storing the streamer level
+struct ParameterDebugStreamer : public o2::conf::ConfigurableParamHelper<ParameterDebugStreamer> {
+  StreamFlags StreamLevel{}; /// flag to store what will be streamed
+  O2ParamDef(ParameterDebugStreamer, "DebugStreamerParam");
+};
+
+#endif
+
+/// class to enable streaming debug information to root files
+class DebugStreamer
+{
+
+// CPU implementation of the class
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE) && defined(DEBUG_STREAMER)
+ public:
+  /// set the streamer i.e. create the output file and set up the streamer
+  /// \param outFile output file name without .root suffix
+  /// \param option RECREATE or UPDATE
+  void setStreamer(const char* outFile, const char* option);
+
+  /// \return returns if the streamer is set
+  bool isStreamerSet() const { return mTreeStreamer ? true : false; }
+
+  /// \return returns streamer object
+  o2::utils::TreeStreamRedirector& getStreamer() { return *mTreeStreamer; }
+
+  /// \return returns streamer level i.e. what will be written to file
+  static StreamFlags getStreamFlags() { return ParameterDebugStreamer::Instance().StreamLevel; }
+
+  ///< return returns unique ID for each CPU thread to give each thread an own output file
+  static size_t getCPUID();
+
+  /// \return returns number of trees in the streamer
+  int getNTrees() const;
+
+  /// \return returns an unique branch name which is not already written in the file
+  /// \param tree name of the tree for which to get a unique tree name
+  std::string getUniqueTreeName(const char* tree) const;
+
+  /// set directly the debug level
+  static void setStreamFlags(const StreamFlags streamFlags) { o2::conf::ConfigurableParam::setValue("DebugStreamerParam", "StreamLevel", static_cast<int>(streamFlags)); }
+
+  /// enable specific streamer flag
+  static void enableStream(const StreamFlags streamFlag);
+
+  /// disable a specific streamer flag
+  static void disableStream(const StreamFlags streamFlag);
+
+  /// check if streamer for specific flag is enabled
+  static bool checkStream(const StreamFlags streamFlag) { return ((getStreamFlags() & streamFlag) == streamFlag); }
+
+  /// merge trees with the same content structure, but different naming
+  /// \param inpFile input file containing several trees with the same content
+  /// \param outFile contains the merged tree from the input file in one branch
+  /// \param option setting which is used for the merging
+  static void mergeTrees(const char* inpFile, const char* outFile, const char* option = "fast");
+
+ private:
+  std::unique_ptr<o2::utils::TreeStreamRedirector> mTreeStreamer; ///< streamer which is used for the debugging
+#else
+
+  // empty implementation of the class for GPU or when the debug streamer is not build for CPU
+ public:
+  /// empty for GPU
+  template <typename... Args>
+  GPUd() void setStreamer(Args... args){};
+
+  /// always false for GPU
+  GPUd() static bool checkStream(const StreamFlags) { return false; }
+
+  class StreamerDummy
+  {
+   public:
+    GPUd() int data() const { return 0; };
+
+    template <typename Type>
+    GPUd() StreamerDummy& operator<<(Type)
+    {
+      return *this;
+    }
+  };
+
+  GPUd() StreamerDummy getStreamer() const { return StreamerDummy{}; };
+
+  template <typename Type>
+  GPUd() StreamerDummy getUniqueTreeName(Type) const
+  {
+    return StreamerDummy{};
+  }
+
+#endif
+};
+
+} // namespace o2::utils
+
+#endif

--- a/Common/Utils/src/CommonUtilsLinkDef.h
+++ b/Common/Utils/src/CommonUtilsLinkDef.h
@@ -39,4 +39,7 @@
 
 #pragma link C++ class o2::utils::IRFrameSelector + ;
 
+#pragma link C++ struct o2::utils::ParameterDebugStreamer + ;
+#pragma link C++ class o2::utils::DebugStreamer + ;
+
 #endif

--- a/Common/Utils/src/DebugStreamer.cxx
+++ b/Common/Utils/src/DebugStreamer.cxx
@@ -1,0 +1,66 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "CommonUtils/DebugStreamer.h"
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE)
+#include <thread>
+#include <fmt/format.h>
+#include "TROOT.h"
+#include "TKey.h"
+#endif
+
+O2ParamImpl(o2::utils::ParameterDebugStreamer);
+
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE) && defined(DEBUG_STREAMER)
+
+void o2::utils::DebugStreamer::setStreamer(const char* outFile, const char* option)
+{
+  if (!isStreamerSet()) {
+    ROOT::EnableThreadSafety();
+    mTreeStreamer = std::make_unique<o2::utils::TreeStreamRedirector>(fmt::format("{}_{}.root", outFile, getCPUID()).data(), option);
+  }
+}
+
+std::string o2::utils::DebugStreamer::getUniqueTreeName(const char* tree) const { return fmt::format("{}_{}", tree, getNTrees()); }
+
+size_t o2::utils::DebugStreamer::getCPUID() { return std::hash<std::thread::id>{}(std::this_thread::get_id()); }
+
+int o2::utils::DebugStreamer::getNTrees() const { return mTreeStreamer->GetFile()->GetListOfKeys()->GetEntries(); }
+
+void o2::utils::DebugStreamer::mergeTrees(const char* inpFile, const char* outFile, const char* option)
+{
+  TFile fInp(inpFile, "READ");
+  TList list;
+  for (TObject* keyAsObj : *fInp.GetListOfKeys()) {
+    const auto key = dynamic_cast<TKey*>(keyAsObj);
+    list.Add((TTree*)fInp.Get(key->GetName()));
+  }
+
+  TFile fOut(outFile, "RECREATE");
+  auto tree = TTree::MergeTrees(&list, option);
+  fOut.WriteObject(tree, "tree");
+}
+
+void o2::utils::DebugStreamer::enableStream(const StreamFlags streamFlag)
+{
+  StreamFlags streamlevel = getStreamFlags();
+  streamlevel = streamFlag | streamlevel;
+  setStreamFlags(streamlevel);
+}
+
+void o2::utils::DebugStreamer::disableStream(const StreamFlags streamFlag)
+{
+  StreamFlags streamlevel = getStreamFlags();
+  streamlevel = (~streamFlag) & streamlevel;
+  setStreamFlags(streamlevel);
+}
+
+#endif

--- a/GPU/GPUTracking/Standalone/CMakeLists.txt
+++ b/GPU/GPUTracking/Standalone/CMakeLists.txt
@@ -164,6 +164,7 @@ include_directories(${GPUTRACKING_DIR}/TPCClusterFinder
                     ${O2_DIR}/Common/Field/include
                     ${O2_DIR}/Common/Constants/include
                     ${O2_DIR}/Common/MathUtils/include
+                    ${O2_DIR}/Common/Utils/include
                     ${O2_DIR}/DataFormats/common/include
                     ${O2_DIR}/DataFormats/Detectors/Common/include
                     ${O2_DIR}/DataFormats/Detectors/ITSMFT/common/include

--- a/GPU/GPUTracking/dEdx/GPUdEdx.h
+++ b/GPU/GPUTracking/dEdx/GPUdEdx.h
@@ -20,6 +20,7 @@
 #include "GPUCommonMath.h"
 #include "GPUParam.h"
 #include "GPUdEdxInfo.h"
+#include "CommonUtils/DebugStreamer.h"
 #if defined(GPUCA_HAVE_O2HEADERS) && !defined(GPUCA_OPENCL1)
 #include "DataFormatsTPC/Defs.h"
 #include "CalibdEdxContainer.h"
@@ -86,6 +87,7 @@ class GPUdEdx
   unsigned char mCount = 0;
   unsigned char mLastROC = 255;
   char mNSubThresh = 0;
+  o2::utils::DebugStreamer mStreamer;
 };
 
 GPUdi() void GPUdEdx::checkSubThresh(int roc)
@@ -181,6 +183,40 @@ GPUdnii() void GPUdEdx::fillCluster(float qtot, float qmax, int padRow, unsigned
   }
   if (qmax < mSubThreshMinMax) {
     mSubThreshMinMax = qmax;
+  }
+
+  using Streamer = o2::utils::DebugStreamer;
+  if (Streamer::checkStream(o2::utils::StreamFlags::streamdEdx)) {
+    mStreamer.setStreamer("debug_dedx", "UPDATE");
+    int regionTmp = region;
+    float absRelPadTmp = absRelPad;
+    float thresholdTmp = threshold;
+    float qMaxTopologyCorrTmp = qMaxTopologyCorr;
+    float qTotTopologyCorrTmp = qTotTopologyCorr;
+    float qMaxResidualCorrTmp = qMaxResidualCorr;
+    float qTotResidualCorrTmp = qTotResidualCorr;
+    float residualGainMapGainTmp = residualGainMapGain;
+    float fullGainMapGainTmp = fullGainMapGain;
+    mStreamer.getStreamer() << mStreamer.getUniqueTreeName("tree").data()
+                            << "qTot=" << mChargeTot[mCount - 1]
+                            << "qMax=" << mChargeMax[mCount - 1]
+                            << "region=" << regionTmp
+                            << "padRow=" << padRow
+                            << "tanTheta=" << tanTheta
+                            << "trackTgl=" << trackTgl
+                            << "sinPhi=" << trackSnp
+                            << "z=" << z
+                            << "absRelPad=" << absRelPadTmp
+                            << "relTime=" << relTime
+                            << "threshold=" << thresholdTmp
+                            << "qTotIn=" << qTotIn
+                            << "qMaxTopologyCorr=" << qMaxTopologyCorrTmp
+                            << "qTotTopologyCorr=" << qTotTopologyCorrTmp
+                            << "qMaxResidualCorr=" << qMaxResidualCorrTmp
+                            << "qTotResidualCorr=" << qTotResidualCorrTmp
+                            << "residualGainMapGain=" << residualGainMapGainTmp
+                            << "fullGainMapGain=" << fullGainMapGainTmp
+                            << "\n";
   }
 }
 

